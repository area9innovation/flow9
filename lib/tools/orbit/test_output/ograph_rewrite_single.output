Original expression: (d * (e + f))
Pattern: (x * (y + z))
Replacement: ((x * y) + (x * z))
Extracted expression ID: 7: (d * (e + f))
Found match at eclass ID: 7
Replacement result: ((2 * 4) + (2 * 5))
Applied 1 rewrites
Rewritten expression: ((2 * 4) + (2 * 5))
Result: ((2 * 4) + (2 * 5))
