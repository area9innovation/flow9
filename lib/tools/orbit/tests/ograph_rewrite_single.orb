// Simplified test for algebraic expression rewriting using orbitRewrite
fn quote(e : ast) = e;

// Test with the distributive property rewrite
// Original expression: d * (e + f)
let expr = quote(d * (e + f));

// Pattern for distributive property: x * (y + z)
let pattern = quote(x * (y + z));

// Replacement: (x * y) + (x * z)
let replacement = quote((x * y) + (x * z));

println("Original expression: " + prettyOrbit(expr));
println("Pattern: " + prettyOrbit(pattern));
println("Replacement: " + prettyOrbit(replacement));

// Create a new OGraph to work with
let graphName = "rewrite_graph";
let graph = makeOGraph(graphName);

// Add the original expression to the graph directly
let exprId = addOGraph(graph, expr);

println("Extracted expression ID: " + i2s(exprId) + ": " + prettyOrbit(extractOGraph(graph, exprId)));

// Apply the rule to the graph with tracing enabled
let matchCount = matchOGraphPattern(graphName, pattern, \(bindings : ast, eclassId) . (
	println("Found match at eclass ID: " + i2s(eclassId));

	// Process the replacement using direct substitution
	let result = substituteWithBindings(replacement, bindings);
	println("Replacement result: " + prettyOrbit(result));
	
	// Add the result to the graph
	let resultId = addOGraph(graph, result);
	
	// Merge the result with the matched node's eclass
	let _ = mergeOGraphNodes(graph, resultId, eclassId);
));

println("Applied " + i2s(matchCount) + " rewrites");

// Extract the rewritten expression
let result = extractOGraph(graph, exprId);

println("Rewritten expression: " + prettyOrbit(result));

result
