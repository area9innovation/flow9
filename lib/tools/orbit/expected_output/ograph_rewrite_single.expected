Original expression: (d * (e + f))
Pattern: (x * (y + z))
Replacement: ((x * y) + (x * z))
Extracted expression ID: 5: (d * (e + f))
Found match at eclass ID: 5
Replacement result: ((1 * 2) + (1 * 3))
Applied 1 rewrites
Rewritten expression: ((1 * 2) + (1 * 3))
Result: ((1 * 2) + (1 * 3))
